### PR TITLE
feat(fiftyone): label LeRobot subtasks and verify exports with YAML

### PR DESCRIPTION
**Pull request by timothy-le7**
## Review follow-up — native FiftyOne export

- Verified the upstream 1.22 exporter: it renumbers episodes in collection order. Directly using source `episode_index` would break subset exports. The fix validates source identities, freezes an explicitly ordered sample view for both mapping and native export, and returns `episode_index_mapping`. Regression coverage includes shuffled/noncontiguous episodes and invalid/duplicate identities.
- Replaced hand-written boto3 upload/URI parsing with shared `StorageClient.upload_directory(require_empty=True)`, using adaptive retries. The SSH/container script bundles the same shared storage implementation so it also works without NPA installed remotely. Tests execute that script with NPA imports blocked.
- **Real FiftyOne 1.22 + MongoDB integration passed:** SO100 episodes imported as `7,2`, exported as `2→0,7→1`; **771/771 real frame identities checked, four labels, zero gaps, four tags restored on native re-import**. This uncovered and fixed a real task-catalog incompatibility: Pandas-index task names are normalized on a separate staging copy during import; original data remains unchanged.
- Fixed the previous Linux CI failures in camera evidence tests: compare exact native YUV planes instead of architecture-dependent RGB conversion hashes. Both 454-frame camera clips were independently matched to the pinned full source. Existing MP4/RRD bytes and receipts remain unchanged and retain a hash-verified historical recipe.
- Merged latest `main`, including the new accepted FiftyOne 1.21 release; retained the distinction between that release and this PR's unpromoted 1.22 source candidate. GitHub reports the branch mergeable. **417 merged-tree targeted tests, 3,599 guardrails, 114 smoke tests, and one real native integration test passed**; Ruff, docs drift, staged confidentiality and gitleaks passed. All 33 accepted public image digests matched anonymous registry reads. These are separate selections. The broader macOS selection had 554 passes and four GNU-coreutils-related failures; using installed GNU tools resolved three, and the remaining SSH staging failure reproduced on clean `main`. The complete Linux merge gate remains required.

[Native validation report](https://github.com/nebius/nebius-physical-ai/blob/codex/lerobot-subtask-label/docs/workbench/evidence/fiftyone-subtasks-native.md) · [Updated labeling guide](https://github.com/nebius/nebius-physical-ai/blob/codex/lerobot-subtask-label/docs/workbench/guides/lerobot-subtask-labeling.md)

Scope: real native API/database export and re-import, with programmatic identity-test tags. Not a new interactive browser-labeling recording or live cloud S3 upload. S3 transport and remote execution are exercised hermetically; no new cloud infrastructure or image was deployed. Fresh CI for this follow-up must be read separately from historical results below.

[Watch the real LeRobot video with eight labels (MP4)](https://github.com/nebius/nebius-physical-ai/raw/refs/heads/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-video-subtasks/lerobot-labeled.mp4) · [Watch its agent UI playback (MP4)](https://github.com/nebius/nebius-physical-ai/raw/refs/heads/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-video-subtasks/agent-ui-playback.mp4)

## Improved: actual robot cameras and multiple labels

The primary proof now uses actual public LeRobot SO-100 footage, not the earlier numeric fixture. Both top and wrist cameras appear together, with a changing active-label badge, all eight subtask labels, a phase timeline, and source frame/time. Labels: **ready → approach → grasp → transfer → place → release → retract → complete**.

- **454/454 actual episode frames, eight labels, zero gaps**; original actions, observations, frame identities, and Arrow types unchanged. The same YAML completed locally and proves `grasp` at episode 0, frame 160, timestamp 5.333333492279053, `subtask_index=2`.
- The direct MP4 is 1280×720, 30 FPS, 15.13 seconds. Its exact bytes are embedded in the new RRD, alongside all 454 labeled source rows and time-aligned video references. The companion MP4 is a real browser screen recording of that RRD playing in the deployed agent UI, not recreated UI frames.
- Both camera clips were remuxed without re-encoding. Independent decoding matched every one of their 454 frames to the pinned upstream files. The composition and browser capture have separate full-decoding and visible-label checks.
- Source: `lerobot/svla_so100_pickplace`, immutable revision `728583b5eaf9e739a7f119e2def466fa1d552402`, Apache-2.0. Source hashes, dataset card, license, changed-file notice, original/derived episode files, annotations, YAML execution report, and reproducible producer are included.
- **Scope:** assistant-authored, visually checked demonstration annotations—not upstream ground truth or expert-certified labels. The production FiftyOne temporal-tag parser/writer is exercised programmatically; this is not a recording of interactive FiftyOne label authoring or a claim of trained-policy performance.

[New evidence and reproduction](https://github.com/nebius/nebius-physical-ai/blob/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-video-subtasks/README.md) · [New RRD](https://github.com/nebius/nebius-physical-ai/raw/refs/heads/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-video-subtasks/lerobot-video-subtasks.rrd)

## Labeling implementation

Connect manual subtask review to the durable LeRobot training dataset. FiftyOne 1.22 imports synchronized LeRobot v3 episodes; reviewed `subtask:<label>` intervals export into a new dataset containing per-frame `subtask_index`, a label catalog, and resumable annotation metadata.

The new `workflows/testing/lerobot-subtask-proof.yaml` validates the exported data and publishes a concrete frame/subtask proof. It checks full coverage and declared dataset counts, resolves indexes through the catalog, and records SHA-256 evidence. It rejects gaps, unknown indexes, missing expected labels, and output locations inside the source dataset. Training code must explicitly consume these labels; this change does not configure hierarchical policy training.

## Workflow and evidence

`LeRobot v3 → FiftyOne timeline review → derived LeRobot export → YAML verification → proof JSON`

The YAML completed locally through `npa workbench workflow run-spec --execute` on a synthetic Parquet fixture:

```json
{"episode_index":0,"frame_index":2,"timestamp":0.2,"subtask_index":2,"subtask":"grasp"}
```

- Coverage: 2 episodes, 8/8 labeled frames, 4 subtask segments, zero unlabeled frames.
- The proof includes source-Parquet, label-catalog, and row hashes. Tests independently recompute them and verify that source bytes remain unchanged.
- CPU live-submit coverage seeds a separate four-frame LeRobot fixture, then reads the published S3 proof and source Parquet/catalog to verify the claimed row and digests. Tampering rejection and run-prefix isolation are tested.
- Operator guide: `docs/workbench/guides/lerobot-subtask-labeling.md`; execution prerequisites: `workflows/testing/lerobot-subtask-proof.readiness.json`.

## Earlier synthetic contract proof (retained)

The MP4 is an actual screen recording of the exact saved RRD in the deployed agent UI's fullscreen embedded Rerun viewer. It steps through all eight rows and holds `grasp` for eight seconds. This is real UI playback of synthetic data, not footage of a robot or FiftyOne editing. The H.264 file is 1680×1120, 25 FPS, 32.92 seconds, and 823 decoded frames. Its companion receipt records video/source hashes, live UI authentication checks, and visual readback at 9 and 24 seconds. No private endpoints, credentials, or infrastructure identifiers are included.

[Open the evidence and viewing instructions](https://github.com/nebius/nebius-physical-ai/blob/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-subtasks/README.md) · [Download the full bundle](https://github.com/nebius/nebius-physical-ai/raw/refs/heads/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-subtasks/proof-bundle.zip) · [Rerun recording](https://github.com/nebius/nebius-physical-ai/raw/refs/heads/codex/lerobot-subtask-label/docs/workbench/evidence/lerobot-subtasks/lerobot-subtasks.rrd)

- The committed bundle includes original and labeled Parquet files, temporal tags, export report, executed YAML/report, all eight before/after rows, and a digest-bound manifest.
- Scrub the Rerun `dataset_row` timeline to **2** to see the original frame beside its derived `grasp` label. Independent decoding verifies all eight samples of each of four entities, not just that the file opens. `rerun rrd verify` reports one file verified without error.
- Every original input file retains its hash; all original frame columns and types are unchanged. The new recording tests also reject mismatched run identities, changed labels, modified source columns, and overwriting an existing bundle.
- Reproduce locally with `PYTHONPATH=npa/src npa/.venv/bin/python npa/scripts/record_lerobot_subtask_proof.py --output-dir results/lerobot-subtasks-new --run-id lerobot-subtasks-new`.
- Scope: **synthetic numeric fixture with programmatic tags**, using the production label parser/writer. This is not real robot footage, a FiftyOne UI capture/native export, or a cloud run.

## Earlier validation (historical)

- Real-video follow-up: **36 focused tests and 3,599 smoke/guardrail tests passed**; full Ruff, staged confidentiality scan, and gitleaks passed. Independent decoding checks both source cameras, all 454 labels, the exact embedded MP4, and all eight visible phases in the 469-frame agent UI capture. No production exporter, verifier, YAML, container, or infrastructure behavior changed in this follow-up; prior macOS full-suite/security limitations below remain applicable, and fresh Linux CI is required.
- MP4 follow-up: **27 focused tests passed**, including full decoding of all 823 video frames, source/video hash binding, and distinct visible playback states; full Ruff passed.
- Recorded proof, YAML execution, and exporter tests: **26 passed**; full Ruff and **114 onboarding smoke tests passed**.
- Follow-up full guardrails: **3,475 passed**. The local security selection reports **663 passed, 67 failed, 8 skipped, 63 teardown errors** on macOS, including the existing Linux `/proc` dependency in isolated-API tests. Its 124 leftover test-owned processes were explicitly identified and stopped; unrelated processes were left untouched. This is not a local security-gate pass; the fresh Linux CI result remains required.
- Before this evidence-only follow-up, all applicable GitHub checks on `edd8d02b68573feb78becfa1e1e06012dd5df402` were green, including Python 3.10/3.12/3.14, browser, security regression, lint, docs, guardrails, and secret/confidentiality scans. The follow-up triggers fresh checks.
- Latest focused exporter, YAML execution, S3 round-trip, live readback, and matrix tests: **193 passed**.
- Workflow/catalog selection: **2,316 passed**; full guardrail selection: **3,469 passed**; onboarding smoke: **114 passed**. These are separate, overlapping selections.
- Workflow schema validation and planning, full Ruff lint, readiness hash, staged/aggregate confidentiality scans, and gitleaks: passed.
- Earlier CLI/image selection: 132 passed. Docs drift and image-tag consistency passed for the labeling implementation.
- Comprehensive macOS suite was interrupted in an unrelated multi-node socket test: **17,095 passed, 149 failed, 63 teardown errors, 97 skipped, 13 deselected, 1 xpassed**. Failures include Linux `/proc` process ownership/cleanup, scanner/build prerequisites, and environment-dependent submit/storage tests. The run's remaining child processes were cleaned up. This is not a full-suite pass; Linux CI remains required.
- Cloud retry: S3 and Nebius authentication both passed, but exact-kubeconfig SkyPilot verification returned `kubernetes_enabled=false`; no cloud workflow was submitted. That earlier cloud run remained gated; the native FiftyOne gap is addressed by the new local integration evidence above.

The fixture establishes the data contract and executable workflow, not real-world label accuracy or policy improvement. The accepted public FiftyOne image pin remains unchanged; the 1.22 source candidate still requires the normal image build, security, functional, and promotion gates.

### CODE REVIEW

**Comment by timothy-le7** [comment]
Merge-readiness follow-up at `7df437fdf5f6e3426ee6689addedffc663edef50`:

- Fixed the Bandit B102 failure in the remote-export regression test: the trusted generated entrypoint now runs from a temporary file with normal `__main__` semantics. The no-installed-NPA and adaptive-storage-retry assertions remain intact; no scanner suppression or policy change was added.
- Merged current `main` (`fa9b7590688af2f047a1309c47c235fedfacb1db`).
- Final-head local validation: **3,662 guardrails passed**, **278 focused/smoke tests passed**, and **1 native FiftyOne 1.22 integration test passed**. The native test checked 771 real LeRobot frames from source episodes 7 and 2, correct derived episode mapping, four labels, zero unlabeled frames, and four restored temporal tags on re-import. Tags were applied programmatically; this is not a claim of a new interactive UI recording.
- Ruff, documentation drift, pinned Bandit, Gitleaks, and the local diff-scoped confidentiality scan passed.
- [The GitHub CI run succeeded](https://github.com/nebius/nebius-physical-ai/actions/runs/35263302153), including all three required contexts: `gitleaks`, `scan`, and `security-regression`. GitHub reports `CLEAN` / `MERGEABLE`.

The full coverage/browser matrix remains a merge-queue gate under repository policy. This PR has not been merged or enqueued.

**Comment by timothy-le7** [comment]
Queue rejection root cause fixed in `e35e6b8f6df6e508cf84d815b4bd098ca7ed5af5`.

The last two queue attempts failed in `test_url_ambiguous_source_names_are_rejected_before_upload`: its preliminary assertion still expected the shared S3 parser to discard query/fragment suffixes. The parser now deliberately rejects them. The test now asserts `StorageError` and retains the archive rejection and zero-upload assertions; no runtime guard or CI policy was weakened.

Both cases reproduced locally before the fix and pass afterward. Additional validation: 72 storage/export tests and 3,918 guardrail/smoke tests passed; Ruff, docs drift, Gitleaks and the diff-scoped confidentiality scan passed. The complete archive restore suite requires Linux `/proc`, so its final validation is the full Linux CI run, not the Mac result.

[Full candidate CI is running](https://github.com/nebius/nebius-physical-ai/actions/runs/35371670818). I will enqueue after the required checks pass; repository auto-merge is disabled.
